### PR TITLE
Better parser

### PR DESCRIPTION
**Pull request by jonludlam**
Let's have a parser that can handle quoted arguments. Additionally, ignore arguments that don't fit the key=value model rather than dying.

### CODE REVIEW

File: lib/parse_argv.ml [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        let rem = String.Sub.tail rem in
+        begin match String.Sub.head rem with
+          | Some c ->
+            let so_far' = String.(sub (of_char c)) :: data :: so_far in
+            inner in_quoted (String.Sub.tail rem) so_far' acc
+          | None -> raise Exit
+        end
+      | Some c ->
+        Printf.printf "Something went wrong in the argv parser: Matched '%c'" c;
+        raise Exit
```
Comment (Drup) [inline lib/parse_argv.ml]: `Bootvar.argv` returns an error type, it would be better to do something similar.

Comment (jonludlam) [inline lib/parse_argv.ml]: That particular case was there just to satisfy the pattern match exhaustiveness check - if we get to that clause, Astring is broken. Having said that, there's an additional 'raise Exit' a few lines above that is reachable, so I should fix that. And if the whole thing is going to return an error type, I might as well fix this one too.

**Comment by jonludlam** [comment]
I cancelled two queued travis builds on older commits, hence the red crosses above.

**Comment by talex5** [comment]
With this branch, booting on Qubes failes with:

```
.[32;1mMirageOS booting....[0m
  start_info: 0000000000273000(VA)
nr_pages: 0x1400
  shared_inf: 0x174bf000(MA)
 pt_base: 0000000000276000(VA)
nr_pt_frames: 0x5
mfn_list: 0000000000269000(VA)
   mod_start: 0x268000(VA)
 mod_len: 4096
   flags: 0x0
cmd_line: root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3 nopat
   stack: 0000000000247900-0000000000267900
MM: Init
  _text: 0000000000000000(VA)
 _etext: 000000000012a870(VA)
   _erodata: 0000000000165000(VA)
 _edata: 000000000020dea0(VA)
stack start: 0000000000247900(VA)
   _end: 0000000000267900(VA)
  start_pfn: 27e
max_pfn: 1400
Mapping memory range 0x400000 - 0x1400000
setting 0000000000000000-0000000000165000 readonly
skipped 1000
MM: Initialise page allocator for 286000(286000)-1400000(1400000)
MM: done
Demand map pfns at 1401000-0000002001401000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 0000000001401000.
Netif: add resume hook
gnttab_stubs.c: initialised mini-os gntmap
kernel args: "root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3 nopat"
Fatal error: exception Failure("Malformed boot parameter \"ro\"")
Raised at file "src/core/lwt.ml", line 901, characters 22-23
Called from file "lib/main.ml", line 58, characters 10-20
Called from file "main.ml", line 52, characters 5-10
Mirage exiting with status 2
Do_exit called!
.[32;1mMirageOS booting....[0m
  start_info: 000000000026e000(VA)
nr_pages: 0x1400
  shared_inf: 0x1fa27000(MA)
 pt_base: 0000000000271000(VA)
nr_pt_frames: 0x5
mfn_list: 0000000000264000(VA)
   mod_start: 0x263000(VA)
 mod_len: 4096
   flags: 0x0
cmd_line: root=/dev/mapper/dmroot ro nomodeset console=hvc0 rd_NO_PLYMOUTH 3 nopat
   stack: 0000000000242b80-0000000000262b80
MM: Init
  _text: 0000000000000000(VA)
 _etext: 0000000000128be0(VA)
   _erodata: 0000000000164000(VA)
 _edata: 0000000000209120(VA)
stack start: 0000000000242b80(VA)
   _end: 0000000000262b80(VA)
  start_pfn: 279
max_pfn: 1400
Mapping memory range 0x400000 - 0x1400000
setting 0000000000000000-0000000000164000 readonly
skipped 1000
MM: Initialise page allocator for 281000(281000)-1400000(1400000)
MM: done
Demand map pfns at 1401000-0000002001401000.
Initialising timer interface
Initialising console ... done.
gnttab_table mapped at 0000000001401000.
Netif: add resume hook
gnttab_stubs.c: initialised mini-os gntmap
qubes-firewall: unknown option `--console'.
Usage: qubes-firewall [OPTION]... 
Try `qubes-firewall --help' for more information.
Ignoring malformed parameter: roIgnoring malformed parameter: nomodesetIgnoring malformed parameter: rd_NO_PLYMOUTHIgnoring malformed parameter: 3Ignoring malformed parameter: nopatMirage exiting with status 1
Do_exit called!
```

I guess it needs some newlines added.

How do I make this work with functoria?

**Comment by jonludlam** [comment]
Ah yes, forgot the newline. Looks like the 'console=hvc0' got through the key=value parser and was turned into '--console hvc0'. We'd need to decide on the approaches outlined in https://github.com/mirage/mirage/issues/493 to progress this I think.

**Comment by dbuenzli** [comment]
What about specifying the environment as pairs of positional arguments ? It removes the need for quoting, see my comment here https://github.com/mirage/mirage-bootvar-xen/issues/6#issuecomment-192228548.

**Comment by hannesm** [comment]
instead of #19, I think this should get merged... ping, anyone around?

**Comment by jonludlam** [comment]
Oops, had forgotten about this. @dbuenzli: This parser is parsing the Xen command line, which is just presented as a single string - your point about the key value pairs, if I understand it correctly, effectively offloads the parsing onto bash by way of it providing your `argv`. In this case, there's no help available, so we still need to have a careful parser.

**Comment by hannesm** [comment]
I've no clue, but shouldn't the mirage-bootvar-xen just expose the string (argv) from xenstore, and this is then passed by mirage/functoria into a cmdliner?  is there any need for string processing here?

**Comment by jonludlam** [comment]
There's definitely a need for _some_ string processing, as Cmdliner needs an array of strings to process, and the Xen PV ABI only provides a single string, so something's got to do some processing. The encoding is done by whoever puts it into the xl.cfg file: 

http://xenbits.xen.org/docs/unstable/man/xl.cfg.5.html#Direct-Kernel-Boot

or by whoever wrote the grub config if pygrub is being used. The equivalent code in the linux kernel is here:

http://lxr.hpcs.cs.tsukuba.ac.jp/linux/kernel/params.c#L171

We could do whatever we like here - a json fragment, binio, whatever. Copying roughly what the linux kernel does seems like a reasonable idea for now.

**Comment by hannesm** [comment]
and this PR even adds some testing... is it good to go, then?

**Comment by samoht** [comment]
LGTM

**Comment by jonludlam** [comment]
In it goes then!

**Comment by hannesm** [comment]
can we have a new release with improved parsing (and modified dependencies)? :)

**Comment by samoht** [comment]
ping @MagnusS

**Comment by hannesm** [comment]
2 weeks passed... any news on a new release? :) (also I guess the open issues should be re-evaluated)

**Comment by yomimono** [comment]
It seems I have rights to cut a release on this repository, so I'll do so soon.

**Comment by MagnusS** [comment]
@yomimono thanks for doing the release!

Sorry for the delay - I have not had time to look at this yet :-/

**Pull request merged**